homeassistant/components/history_stats/sensor.py

```diff
@@ -124,35 +124,40 @@ def __init__(
         self._name = name
         self._unit_of_measurement = UNITS[sensor_type]
 
-        self._period = (datetime.datetime.now(), datetime.datetime.now())
+        self._period = (datetime.datetime.min, datetime.datetime.min)
         self.value = None
         self.count = None
+        self._history_current_period: list[State] = []
+        self._previous_run_before_start = False
+
+    @callback
+    def _async_start_refresh(self, *_) -> None:
+        """Register state tracking."""
+        self.async_schedule_update_ha_state(True)
+        self.async_on_remove(
+            async_track_state_change_event(
+                self.hass, [self._entity_id], self._async_update_from_event
+            )
+        )
 
     async def async_added_to_hass(self):
         """Create listeners when the entity is added."""
-
-        @callback
-        def start_refresh(*args):
-            """Register state tracking."""
-
-            @callback
-            def force_refresh(*args):
-                """Force the component to refresh."""
-                self.async_schedule_update_ha_state(True)
-
-            force_refresh()
-            self.async_on_remove(
-                async_track_state_change_event(
-                    self.hass, [self._entity_id], force_refresh
-                )
-            )
-
         if self.hass.state == CoreState.running:
-            start_refresh()
+            self._async_start_refresh()
             return
-
         # Delay first refresh to keep startup fast
-        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, start_refresh)
+        self.hass.bus.async_listen_once(
+            EVENT_HOMEASSISTANT_START, self._async_start_refresh
+        )
+
+    async def async_update(self) -> None:
+        """Get the latest data and updates the states."""
+        await self._async_update(None)
+
```

Comment:
Should this method also call `self.async_write_ha_state()` ?

Reply:
This one is sadly still doing polling to update.  I want to switch it to use DataUpdateCoordinator and extract the complexity out of the entity class in a future PR